commit 95dbe336d30e9356c20fc37f6351a0d0c17ad790
Author: dann frazier <dann.frazier@canonical.com>
Date:   Thu Nov 18 15:03:09 2010 -0700

    ocfs2_connection_find() returns pointer to bad structure
    
    commit 226291aa4641fa13cb5dec3bcb3379faa83009e2 upstream.
    
    If ocfs2_live_connection_list is empty, ocfs2_connection_find() will return
    a pointer to the LIST_HEAD, cast as a ocfs2_live_connection. This can cause
    an oops when ocfs2_control_send_down() dereferences c->oc_conn:
    
    Call Trace:
      [<ffffffffa00c2a3c>] ocfs2_control_message+0x28c/0x2b0 [ocfs2_stack_user]
      [<ffffffffa00c2a95>] ocfs2_control_write+0x35/0xb0 [ocfs2_stack_user]
      [<ffffffff81143a88>] vfs_write+0xb8/0x1a0
      [<ffffffff8155cc13>] ? do_page_fault+0x153/0x3b0
      [<ffffffff811442f1>] sys_write+0x51/0x80
      [<ffffffff810121b2>] system_call_fastpath+0x16/0x1b
    
    Fix by explicitly returning NULL if no match is found.
    
    Signed-off-by: dann frazier <dann.frazier@canonical.com>
    Signed-off-by: Joel Becker <joel.becker@oracle.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

diff --git a/fs/ocfs2/stack_user.c b/fs/ocfs2/stack_user.c
index 353fc35..2179c7d 100644
--- a/fs/ocfs2/stack_user.c
+++ b/fs/ocfs2/stack_user.c
@@ -190,7 +190,7 @@ static struct ocfs2_live_connection *ocfs2_connection_find(const char *name)
 			return c;
 	}
 
-	return c;
+	return NULL;
 }
 
 /*
